src/dbManager.js

```diff
@@ -0,0 +1,161 @@
+const BN = require('bn.js')
+const level = require('level-mem')
+const rlp = require('rlp')
+const Block = require('ethereumjs-block')
+const Cache = require('./cache')
+const {
+  headsKey,
+  headHeaderKey,
+  headBlockKey,
+  hashToNumberKey,
+  numberToHashKey,
+  tdKey,
+  bodyKey,
+  headerKey
+} = require('./util')
+
+module.exports = class DBManager {
+  constructor (db, common) {
+    this._db = db
+    this._common = common
+    this._opts = {
+      keyEncoding: 'binary',
+      valueEncoding: 'binary'
+    }
+    this._cache = {
+      td: new Cache({ max: 1024 }),
+      header: new Cache({ max: 512 }),
+      body: new Cache({ max: 256 }),
+      numberToHash: new Cache({ max: 2048 }),
+      hashToNumber: new Cache({ max: 2048 })
+    }
+  }
+
+  getHeads () {
+    return this._db.get(headsKey, {
+      keyEncoding: 'binary',
+      valueEncoding: 'json'
+    })
+  }
+
+  getHeadHeader () {
+    return this._db.get(headHeaderKey, this._opts)
+  }
+
+  getHeadBlock () {
+    return this._db.get(headBlockKey, this._opts)
+  }
+
+  get (k) {
+    return this._db.get(k, this._opts)
+  }
+
+  batch (ops) {
+    return this._db.batch(ops)
+  }
+
+  /**
+   * Performs a block hash to block number lookup
+   * @method _hashToNumber
+   */
+  async _hashToNumber (hash) {
+    const key = hashToNumberKey(hash)
+    let number = this._cache.hashToNumber.get(key)
+    if (number) {
+      return new BN(number)
+    }
+
+    number = await this.get(key)
+    this._cache.hashToNumber.set(key, number)
+
+    return new BN(number)
+  }
+
+  /**
+   * Performs a block number to block hash lookup
+   * @method _numberToHash
+   */
+  async _numberToHash (number) {
+    if (number.ltn(0)) {
+      throw new level.errors.NotFoundError()
+    }
+
+    const key = numberToHashKey(number)
+    let hash = this._cache.numberToHash.get(key)
+    if (hash) {
+      return hash
+    }
+
+    hash = await this._db.get(key)
+    this._cache.numberToHash.set(key, hash)
+
+    return hash
+  }
+
```

Comment:
```suggestion
  async _numberToHash (number) {
    if (number.ltn(0)) {
      throw new level.errors.NotFoundError()
    }
    return await this.get(numberToHashKey(number), {cache: 'numberToHash'})
  }
```

Reply:
@s1na Is the `async`  and `await` not necessary here?

Reply:
I think in this case both alternatives (having `async/await` or not) are equivalent. `async` is necessary when there's an `await` in the function, but because `this.get` already returns a `Promise` I think it's safe to just return its result. I could add the `async` and `await` if you think it helps with consistency and readability?

Reply:
Yes, maybe add. Think it helps on readability, probably many first-rough-readers like me know would otherwise stumble upon this.

Reply:
Just tried to add it but `standard` complained: `Redundant use of `await` on a return value`

Reply:
Ok, then just leave.